### ISSUE TITLE
fix wildcard search in new app

### ISSUE THREAD

**Issue by TuukkaOT** [kanerv/TuukkaSaanaKanerva#29]
I added the try except keyerror back because it was a problem when trying to fix wildcard searches.
EDIT: only works with words that have no hyphenated or otherwise punctuated forms.

**Comment by kanerv** [comment]
yes, the hyphenated-words are not working even with the exact match... I wonder if we should just get rid of all hyphens when processing the data and search terms. However, we would need to somehow get them back for printing the results 😕

**Comment by TuukkaOT** [comment]
I think the problem might be with TfidVectorizers default tokenizer that tokenizes hyphenated words as separate words. It can be changed by putting token_pattern=r"" with a regex inside the quoatation marks as a parametre in the vectorizer. I managed to get it to not complain about hyphenated words but it still had a problem with apostrophes in genitives. I might be wrong here but that's what I was able to came up with.

**Comment by kanerv** [comment]
> I think the problem might be with TfidVectorizers default tokenizer that tokenizes hyphenated words as separate words. It can be changed by putting token_pattern=r"" with a regex inside the quoatation marks as a parametre in the vectorizer. I managed to get it to not complain about hyphenated words but it still had a problem with apostrophes in genitives. I might be wrong here but that's what I was able to came up with.

Sounds convincing to me!

**Comment by kanerv** [comment]
I also realised that the wildcard search does only print results for one of the searches and not the whole list as our last week's project did. So that needs some more work still. I have a driving test tomorrow, which has been stealing my focus for the past week, but I'll try to get to this project tomorrow as well.

**Comment by saanahyttinen** [comment]
I was thinking if this could be fixed with a try-except:KeyError, because what I get when searching for example house* is a KeyError for 'house-holders'. I tried to enter some of the test_query to a try-except structure but that didn't help at least...

**Comment by kanerv** [comment]
I worked on the hyphenated words today with Raul and now the words including "**'**" (like _one's_) or "**-**" (like _house-holders_) are tagged as single units and not split as they were previously.

**Comment by TuukkaOT** [comment]
I added . and " because there were some cases where those were blocking our wildcard search. I don't know what happened but now the results for the wildcard search come as stemmed words and the search is suuper slow. I will probably have time to check this out tomorrow. If I messed up something, feel free to backtrack or take away the try except and regex changes.

**Comment by kanerv** [comment]
I tried building a function (kinda like the _stem()_-function) that would find all the words matching the wildcard query in the data and change them to _word_wc_ (kinda like the stemmed words have the __s_ notation). This way we could search for all of them with a query like _query_wc_. However, I could not get the function to work. I might have just a little mistake when appending new words to document_out **or** there is something bigger going wrong... Feel free to experiment if you want to! 🚽

**Comment by TuukkaOT** [comment]
I can't seem to able to crack this either. I noticed one thing that the split() in the wildcard function leaves some extra parentheses in the text that python has a problem with. Maybe we could change the split to a tokenizer like in stem() function. Also I think the reason why the app is so slow might be because it searches every wildcard separately. But I don't know if there's anything we can do about that.

**Comment by kanerv** [comment]
I fixed it!!! Now there is a separate function _test_wcquery(query)_ for searching wildcards. I had to look at another groups' code to understand how to do it, but I credited them in the comments.

The wildcards are no longer searched separately (which is how it should have been from the start and therefore it's a lot quicker as well). First, the function finds all possible words from the documents to match the wildcard query. Then, the program creates a vector matching all of them, so the wildcards are all ranked in the same search.

I've only done some basic testing, but so far it should work. Let me know if you find bugs! 🐛
